jetty-core/jetty-server/src/main/java/org/eclipse/jetty/server/handler/DoSHandler.java

```diff
@@ -0,0 +1,622 @@
+        }
+
+        @Override
+        public long getExpireNanoTime()
+        {
+            return _expireAt;
+        }
+
+        @Override
+        public String toString()
+        {
+            try (AutoLock l = _lock.lock())
+            {
+                return "Tracker@%s{rc=%s}".formatted(_id, _rateControl);
+            }
+        }
+    }
+
+    /**
+     * A {@link RateControl.Factory} that uses an
+     * <a href="https://en.wikipedia.org/wiki/Moving_average#Exponential_moving_average">Exponential Moving Average</a>
+     * to limit the request rate to a maximum number of requests per second.
+     */
+    public static class ExponentialMovingAverageRateControlFactory implements RateControl.Factory
+    {
+        private final Duration _samplePeriod;
+        private final double _alpha;
+        private final int _maxRequestsPerSecond;
+
+        public ExponentialMovingAverageRateControlFactory()
+        {
+            this(null, -1.0, 1000);
+        }
+
+        public ExponentialMovingAverageRateControlFactory(@Name("maxRequestsPerSecond") int maxRateRequestsPerSecond)
+        {
+            this(null, -1.0, maxRateRequestsPerSecond);
+        }
+
+        public ExponentialMovingAverageRateControlFactory(
+            @Name("samplePeriodMs") long samplePeriodMs,
+            @Name("alpha") double alpha,
+            @Name("maxRequestsPerSecond") int maxRequestsPerSecond)
+        {
+            this(samplePeriodMs <= 0 ? null : Duration.ofMillis(samplePeriodMs), alpha, maxRequestsPerSecond);
+        }
+
+        public ExponentialMovingAverageRateControlFactory(
+            @Name("samplePeriod") Duration samplePeriod,
+            @Name("alpha") double alpha,
+            @Name("maxRequestsPerSecond") int maxRequestsPerSecond)
+        {
+            _samplePeriod = samplePeriod == null ? Duration.ofMillis(100) : samplePeriod;
+            _alpha = alpha <= 0.0 ? 0.2 : alpha;
+            if (_samplePeriod.compareTo(Duration.ofSeconds(1)) > 0)
+                throw new IllegalArgumentException("Sample period must be less than or equal to 1 second");
+            if (_alpha > 1.0)
+                throw new IllegalArgumentException("Alpha " + _alpha + " is too large");
+            _maxRequestsPerSecond = maxRequestsPerSecond;
+        }
+
+        @Override
+        public RateControl newRateControl()
+        {
+            return new ExponentialMovingAverageRateControl();
+        }
+
+        class ExponentialMovingAverageRateControl implements RateControl
+        {
+            private double _exponentialMovingAverage;
+            private int _sampleCount;
+            private long _sampleStart;
+
+            private ExponentialMovingAverageRateControl()
+            {
+                _sampleStart = NanoTime.now();
+            }
+
+            @Override
+            public boolean isRateExceededBySample(long now)
+            {
+                // Count the request
+                _sampleCount++;
```

Comment:
I'd use atomics and remove the need for a lock in `Tracker`.

Reply:
You cannot use atomicS plural without it being a race.  There is too much info to record into a single atomic long.   We could use an atomicReference, but then we'd be allocating... maybe that is better than a lock? @lorban???

Let's get the rest of the class agreed first and we can consider a better lock